gh-146636: PEP 803: Reference documentation (#148013)
python/cpython: Doc/c-api/stable.rst

Projects that use this API are expected to follow
CPython development and spend extra effort adjusting to changes.

.. _stable-abi:
.. _stable-application-binary-interface:

Stable Application Binary Interfaces
====================================

Python's :dfn:`Stable ABI` allows extensions to be compatible with multiple
versions of Python, without recompilation.

.. note::

   For simplicity, this document talks about *extensions*, but Stable ABI
   works the same way for all uses of the API – for example, embedding Python.

There are two Stable ABIs:

> - ``abi3``, introduced in Pyton 3.2, is compatible with

Reviewer: ```suggestion
- ``abi3``, introduced in Python 3.2, is compatible with
```

>   **non**-:term:`free threaded <free-threaded build>` builds of CPython.

Reviewer 2: Nitpicky, but I believe "free threaded" needs a dash here. In the original issue where we added the "free-threaded build" term (#144277), we determined that no dash was correct only when "free threaded" was a noun.

> - ``abi3t``, introduced in Pyton 3.15, is compatible with

Reviewer: ```suggestion
- ``abi3t``, introduced in Python 3.15, is compatible with
```

   .. versionadded:: next

      ``abi3t`` was added in :pep:`803`

It is possible for an extension to be compiled for *both* ``abi3`` and
``abi3t`` at the same time; the result will be compatible with
both free-threaded and non-free-threaded builds of Python.
Currently, this has no downsides compared to compiling for ``abi3t`` only.

Each Stable ABI is versioned using the first two numbers of the Python version.
For example, Stable ABI 3.14 corresponds to Python 3.14.
An extension compiled for Stable ABI 3.x is ABI-compatible with Python 3.x
and above.

Extensions that target a stable ABI must only use a limited subset of
the C API. This subset is known as the :dfn:`Limited API`; its contents
are :ref:`listed below <limited-api-list>`.

On Windows, extensions that use a Stable ABI should be linked against
``python3.dll`` rather than a version-specific library such as
``python39.dll``.
This library only exposes the relevant symbols.

On some platforms, Python will look for and load shared library files named
with the ``abi3`` or ``abi3t`` tag (for example, ``mymodule.abi3.so``).
:term:`Free threaded <free-threaded build>` interpreters only recognize the

> ``abi3t`` tag, while non-free threaded ones will prefer ``abi3`` but fall back

Reviewer: ```suggestion
``abi3t`` tag, while non-free-threaded ones will prefer ``abi3`` but fall back
```

to ``abi3t``.
Thus, extensions compatible with both ABIs should use the ``abi3t`` tag.

Python does not necessarily check that extensions it loads
have compatible ABI.
Extension authors are encouraged to add a check using the :c:macro:`Py_mod_abi`
slot or the :c:func:`PyABIInfo_Check` function, but the user
(or their packaging tool) is ultimately responsible for ensuring that,
for example, extensions built for Stable ABI 3.10 are not installed for lower
versions of Python.

All functions in Stable ABI are present as functions in Python's shared
library, not solely as macros.
This makes them usable are usable from languages that don't use the C
preprocessor, including Python's :py:mod:`ctypes`.

.. _abi3-compiling:

Compiling for Stable ABI
------------------------

.. note::

>    When using a build tool (for example, ``setuptools``), the tool is

Reviewer: ```suggestion
   When using a build tool (for example, Setuptools), the tool is
```

Stable ABI Scope and Performance
--------------------------------

The goal for Stable ABI is to allow everything that is possible with the
full C API, but possibly with a performance penalty.
Generally, compatibility with Stable ABI will require some changes to an
extension's source code.

For example, while :c:func:`PyList_GetItem` is available, its "unsafe" macro
variant :c:func:`PyList_GET_ITEM` is not.
The macro can be faster because it can rely on version-specific implementation
details of the list object.

For another example, when *not* compiling for Stable ABI, some C API
functions are inlined or replaced by macros.
Compiling for Stable ABI disables this inlining, allowing stability as
Python's data structures are improved, but possibly reducing performance.

By leaving out the :c:macro:`!Py_LIMITED_API` or :c:macro:`!Py_TARGET_ABI3T`
definition, it is possible to compile Stable-ABI-compatible source
for a version-specific ABI.
A potentially faster version-specific extension can then be distributed
alongside a version compiled for Stable ABI -- a slower but more compatible
fallback.

Stable ABI Caveats
------------------

> Note that compiling for Stable ABI is *not* a complete guarantee that code will
> be compatible with the expected Python versions.
> Stable ABI prevents *ABI* issues, like linker errors due to missing

Reviewer 2: I think we need to precede "Stable ABI" with "the"?

```suggestion
Note that compiling for the Stable ABI is *not* a complete guarantee that code will
be compatible with the expected Python versions.
The Stable ABI prevents *ABI* issues, like linker errors due to missing
```

Author: Even if there's two of them?

Reviewer 2: Yeah, not too sure about the rules here. What about "the Stable ABIs"?

Author: The way I wrote it still sounds best to me...

symbols or data corruption due to changes in structure layouts or function
signatures.
However, other changes in Python can change the *behavior* of extensions.

> See Python's Backwards Compatibility Policy (:pep:`387`) for details.

Reviewer 2: This sentence confused me a little when I first read it. I had read it as "We might change the behavior of extensions -- go read our backward compatibility policy if you want to know what", but then we provide an example right after the reader was diverted. Could we move this sentence to after the following paragraph, or put it in a `seealso` note?

Author: PEP 387 is already mentioned on this page. I removed this reference.

with the ``abi3`` or ``abi3t`` tag (for example, ``mymodule.abi3.so``).
:term:`Free-threaded <free-threaded build>` interpreters only recognize the
``abi3t`` tag, while non-free-threaded ones will prefer ``abi3`` but fall back
to ``abi3t``.
Thus, extensions compatible with both ABIs should use the ``abi3t`` tag.

Python does not necessarily check that extensions it loads
have compatible ABI.
Extension authors are encouraged to add a check using the :c:macro:`Py_mod_abi`
slot or the :c:func:`PyABIInfo_Check` function, but the user
(or their packaging tool) is ultimately responsible for ensuring that,
for example, extensions built for Stable ABI 3.10 are not installed for lower
versions of Python.

All functions in Stable ABI are present as functions in Python's shared
library, not solely as macros.
This makes them usable are usable from languages that don't use the C
preprocessor, including Python's :py:mod:`ctypes`.

.. _abi3-compiling:

Compiling for Stable ABI
------------------------

.. note::

   When using a build tool (for example, ``setuptools``), the tool is
   generally responsible for setting macros and synchronizing them with
   extension filenames and other metadata.

>    Prefer using the tool's options over defining the macros manually.

Reviewer 3: Maybe setuptools is a poor example to use above then? Because setuptools expects users to set the build flags manually...

You could delete the "(For example, SetupTools)" above, or maybe refer to meson-python or scikit-build-core instead? These days that's what I recommend to people depending on if they already use CMake or not.

Author: AFAIK Setuptools is the most widely *known* build tool; the example is there to clarify what kind of “build tool” I'm talking about.

Setuptools has a `--py_limited_api` argument. All tools will ideally need an update for `abi3t`.

   The rest of this section is mainly relevant for tool authors, and for
   people who compile extensions manually.

   .. seealso:: `list of recommended tools`_ in the Python Packaging User Guide

      .. _list of recommended tools: https://packaging.python.org/en/latest/guides/tool-recommendations/#build-backends-for-extension-modules

To compile for a Stable ABI, define one or both of the following macros
to the lowest Python version your extension should support, in
:c:macro:`Py_PACK_VERSION` format.
Typically, you should choose a specific value rather than the version of
the Python headers you are compiling against.

> The macro(s) must be defined before including ``Python.h``.

Reviewer: "(s)" is a bit awkward, prefer a generic singular or plural:

```suggestion
The macros must be defined before including ``Python.h``.
```

Since :c:macro:`Py_PACK_VERSION` is not available at this point, you
will need to use the numeric value directly.
For reference, the values for a few recent Python versions are:

.. version-hex-cheatsheet::

> When the macro(s) are defined, ``Python.h`` will only expose API that is

Reviewer: ```suggestion
When the macros are defined, ``Python.h`` will only expose API that is
```

compatible with the given Stable ABI -- that is, the
:ref:`Limited API <limited-api-list>` plus some definitions that need to be
visible to the compiler but should not be used directly.

.. c:macro:: Py_LIMITED_API

   Target ``abi3``, that is,

>    non-:term:`free-threaded <free-threaded build>` builds of CPython.

Reviewer: ```suggestion
   non-:term:`free-threaded builds <free-threaded build>` of CPython.
```

   See :ref:`above <abi3-compiling>` for common information.

.. c:macro:: Py_TARGET_ABI3T

   Target ``abi3t``, that is,

>    :term:`free-threaded <free-threaded build>` builds of CPython.

Reviewer: ```suggestion
   :term:`free-threaded builds <free-threaded build>` of CPython.
```

>    See :ref:`above <abi3-compiling>` for common information.

Reviewer 3: Maybe add "The free-threaded and GIL-enabled can both load abi3t extensions".

Author: That's not the case.

   .. versionadded:: next

Both macros specify a target ABI; the different naming style is due to
backwards compatibility.

.. admonition:: Historical note

   You can also define ``Py_LIMITED_API`` as ``3``. This works the same as
   ``0x03020000`` (Python 3.2, the version that introduced Stable ABI).

When both are defined, ``Python.h`` may, or may not, redefine
:c:macro:`!Py_LIMITED_API` to match :c:macro:`!Py_TARGET_ABI3T`.

> On a a :term:`free-threaded build` -- that is, when

Reviewer: ```suggestion
On a :term:`free-threaded build` -- that is, when
```

:c:macro:`Py_GIL_DISABLED` is defined -- :c:macro:`!Py_TARGET_ABI3T`
defaults to the value of :c:macro:`!Py_TARGET_ABI3T`.
This means that there are two ways to build for both ``abi3`` and ``abi3t``:

- define both :c:macro:`!Py_LIMITED_API` and :c:macro:`!Py_TARGET_ABI3T`, or

> - define only :c:macro:`!Py_LIMITED_API` and build for free-threaded Python.

Reviewer 3: ```suggestion
- Define only :c:macro:`!Py_LIMITED_API` and define :c:macro:`!Py_GIL_DISABLED`. 
  This happens automatically with a free-threaded interpreter, except on Windows.
```

Author: I think defining Py_GIL_DISABLED "manually" is still only supported on Windows.
